### PR TITLE
[LANG-1692] Cast FieldUtils.readField result to the recipient type

### PR DESCRIPTION
**Pull request by tisonkun**
This patch should be easy to understand.

Generally, it won't affect user cases as long as the type can be inferred by the recipient or JDK rules. There's a case as shown in `ReflectionDiffBuilder.java` that if the type inference results in ambiguity, it may cause compile error.

Thus, I'm unsure whether this is proper to be included with a minor version bump. I don't think adding another series of methods helps - it causes further confusion.

This patch doesn't cause extra class cast exception because if the recipient is `Object` the cast should always succeed. And if the recipient already cast it into another type, there will be an exception if the cast cannot be performed.

### CODE REVIEW

File: src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java [hunk 1]
```diff
@@ -110,8 +110,9 @@ private void appendFields(final Class<?> clazz) {
         for (final Field field : FieldUtils.getAllFields(clazz)) {
             if (accept(field)) {
                 try {
-                    diffBuilder.append(field.getName(), readField(field, left, true),
-                            readField(field, right, true));
+                    Object leftObject = readField(field, left, true);
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:113]: Is this change needed for the type changes? Please revert it otherwise.

**Comment by garydgregory** [comment]
This PR has no matching tests.

File: src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java (hunk 1 above)
Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:113]: Please read the description:

> There's a case as shown in `ReflectionDiffBuilder.java` that if the type inference results in ambiguity, it may cause compile error.

**Comment by tisonkun** [comment]
> This PR has no matching tests.

Please read the description:

> * it won't affect user cases as long as the type can be inferred by the recipient or JDK rules
> * This patch doesn't cause extra class cast exception ...

No existing tests break is fine. I don't know how to add a new effective test.

**Comment by tisonkun** [comment]
I may write something like:

```diff
diff --git a/src/test/java/org/apache/commons/lang3/reflect/FieldUtilsTest.java b/src/test/java/org/apache/commons/lang3/reflect/FieldUtilsTest.java
index b68a994f3..dcecbca06 100644
--- a/src/test/java/org/apache/commons/lang3/reflect/FieldUtilsTest.java
+++ b/src/test/java/org/apache/commons/lang3/reflect/FieldUtilsTest.java
@@ -480,7 +480,8 @@ public void testReadDeclaredNamedStaticFieldForceAccess() throws Exception {
     @Test
     public void testReadField() throws Exception {
         final Field parentS = FieldUtils.getDeclaredField(parentClass, "s");
-        assertEquals("s", FieldUtils.readField(parentS, publicChild));
+        String s = FieldUtils.readField(parentS, publicChild);
+        assertEquals("s", s);
         assertEquals("s", FieldUtils.readField(parentS, publiclyShadowedChild));
         assertEquals("s", FieldUtils.readField(parentS, privatelyShadowedChild));
         final Field parentB = FieldUtils.getDeclaredField(parentClass, "b", true);
```

to prove this change works. But it seems useless.

**Comment by garydgregory** [comment]
You need to think about how to avoid regressions in case the change is reverted. IOW, in this case, you should get a compilation error.

**Comment by tisonkun** [comment]
@garydgregory Thanks for your comments. Will push a followup later this week.

File: src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java (hunk 1 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:113]: @tisonkun 
I still don't get it, why is this needed in this PR as it relates to the generics change?

File: src/main/java/org/apache/commons/lang3/reflect/FieldUtils.java [hunk 2]
```diff
@@ -254,23 +254,23 @@ public static List<Field> getFieldsListWithAnnotation(final Class<?> cls, final
     /**
      * Reads an accessible {@code static} {@link Field}.
      *
-     * @param field
```
[the remark this answers is not in the document]
Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/reflect/FieldUtils.java:257]: Reverting...

[the remark this answers is not in the document]
Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/reflect/FieldUtils.java:257]: Oops. I think I should add the `@param <T>` line so adjusting neighbor lines should be OK. Said you're updating the comments, and adjusting a bit the layout should not be a big deal. I don't touch unrelated comments.

**Comment by tisonkun** [comment]
Tests added.

File: src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java (hunk 1 above)
Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:113]: image

Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:113]:
```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/chenzili/Brittani/commons-lang/src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:[113,32] reference to append is ambiguous
  both method append(java.lang.String,short[],short[]) in org.apache.commons.lang3.builder.DiffBuilder and method append(java.lang.String,java.lang.Object[],java.lang.Object[]) in org.apache.commons.lang3.builder.DiffBuilder match
```

**Comment by tisonkun** [comment]
@garydgregory is there any remaining blocker for this patch?

**Comment by garydgregory** [comment]
I'll take a look this weekend...

File: src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java (hunk 1 above)
Comment (garydgregory) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:113]: Revert this change, please.

Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java:113]: Revert this change will cause compile to fail. I can do it and you see what happens.

Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java]: Reverted at bc4e02229407a80b1315c8c5f00cad859724aa4e.

Now the compilation failed: https://github.com/apache/commons-lang/actions/runs/3210472599/jobs/5247938888.

Comment (garydgregory) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java]: Does that mean that existing call sites in users' applications will also fail?

Comment (tisonkun) [reply src/main/java/org/apache/commons/lang3/builder/ReflectionDiffBuilder.java]: If you think it breaks the existing code and should not do it, you can reject the PR. But I'm a bit confused since reverting it obviously causes compile to fail.

**Comment by tisonkun** [comment]
Closed as no response. I'll add wrappers for this changes downstream when necessary.
